### PR TITLE
[WIP] Serial interface for platform bus w/ listen mode

### PR DESCRIPTION
**Pull request by forGGe**

### CODE REVIEW

File: dev/bus/export/dev/serial.hpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            const auto read_sz = std::min(sz - available_size_1,
+                    available_size_2);
+            std::copy(input_buffer_, input_buffer_ + read_sz,
+                    buf + available_size_1);
+        }
+    }
+    input_read_end_ += sz;
+    auto s = config::in_capacity;
+    input_read_end_.compare_exchange_weak(s, 0);
+    schedule_xfer();
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Why to schedule xfer every time?

Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Avoid any implementation that will trigger xfer every time

File: dev/bus/export/dev/serial.hpp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                // TODO: smarter error handling?
+                return;
+            }
+            // It is not possible to half-transfer one byte
+            ecl_assert(type != bus_event::ht);
+            input_write_end_++;
+
+            // Transfer has been finished
+            // Schedule new xfer if possible
+            if (total == input_current_read_size_) {
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Use binary semaphore `some data present` (0 when there is no data, 1 if there is some data). Set it from ISR, wait in `get_*` functions

File: dev/bus/export/dev/serial.hpp [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                // Wrap the write end
+                auto sz = config::in_capacity;
+                input_write_end_.compare_exchange_weak(sz, 0);
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Why use CE here? ISR will be executed with priority higher than main execution thread

File: dev/bus/export/dev/serial.hpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                schedule_xfer(); // TODO: handle xfer errors
+            }
+    }
+    // TODO: handle TX events
+}
+
+template <class PBus, class SerialConfig>
+err serial<PBus, SerialConfig>::schedule_xfer()
+{
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: It is possible to simplify it. Why to schedule xfer with just part of the buffer? It is possible to use whole buffer every time xfer ends.

File: dev/bus/export/dev/serial.hpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        // are equal, and the entire buffer is available
+        input_current_read_size_ = config::in_capacity;
+    } else {
+        input_current_read_size_ = input_write_end_ > input_read_end_ ?
+            config::in_capacity - input_write_end_ :
+            input_read_end_ - input_write_end_;
+    }
+    // No space in buffer, defer xfer until next time
+    if (input_current_read_size_ == 0) {
+        xfer_state_ = xfer_deferred_;
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: I propose not to keep old data, but rather re-write it with new one. It means that new xfer started every time old is finished.

Now, you can create a flag, meaning `reading in progress` that will signify that buffer is in use. If so, xfer must be started from main thread when it will finish its work, rather than from ISR thread.

File: dev/bus/export/dev/serial.hpp [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    static constexpr auto out_capacity = 128;
+};
+
+//! Serial driver interface.
+//! \details The serial allows to abstract async,
+//! interrupt-driven nature of platform-level drivers and
+//! provide synchronus, buffered data management interface.
+//! \tparam PBus Exclusevely owned platform bus.
+template<class PBus, class SerialConfig = serial_default_config>
+class serial
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Expose buffer size and underlying type to serials' clients:
```
static constexpr auto buffer_size =  ... ;
using platform_handle = PBus;
```

File: dev/bus/export/dev/serial.hpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    return recv_buf(&byte, sz);
+}
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::recv_buf(uint8_t *buf, size_t &sz)
+{
+    ecl_assert(m_is_inited == true);
+    // Restrict starting new xfer from bus_handler()
+    // to prevent data loss
+    m_rx_new_xfer_allowed.wait();
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Probably, you can use atomic here instead of semaphore taking to the account ISR priority. It will be faster on targets with OS enabled.

File: dev/bus/export/dev/serial.hpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+    // Wait until data is available on the read end
+    while ((m_rx_read_iter == m_rx_write_iter)) { }
+
+    sz = std::min(sz, m_rx_write_iter - m_rx_read_iter);
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: `m_rx_write_iter` can be modified during execution of this code. Copy it onto stack to avoid it

File: dev/bus/export/dev/serial.hpp [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+err serial<PBus, buf_size>::try_start_xfer()
+{
+    err result = err::ok;
+    if (m_rx_write_iter == buffer_size) {
+        m_rx_write_iter = 0;
+        // Notify caller that some unread data might be lost
+        if (m_rx_read_iter < buffer_size) {
+            result = err::nobufs;
+        }
+        m_rx_read_iter = 0;
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Suppose that `m_rx_read_iter = buffer_size /  2` and `m_rx_write_iter == buffer_size`. If so, half of the buffer will be lost.

File: dev/bus/export/dev/serial.hpp (hunk 8 above)
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Also, considering comments below:

```
bytes_present = write_pos >= read_pos ?  write_pos - read_pos : buffer_size - read_pos
```

File: dev/bus/export/dev/serial.hpp [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::recv_buf(uint8_t *buf, size_t &sz)
+{
+    ecl_assert(m_is_inited == true);
+    // Restrict starting new xfer from bus_handler()
+    // to prevent data loss
+    m_rx_new_xfer_allowed.wait();
+
+    // Wait until data is available on the read end
+    while ((m_rx_read_iter == m_rx_write_iter)) { }
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: But here semaphore is better.

File: dev/bus/tests/serial_unit.cpp [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+    void teardown()
+    {
+        mock().clear();
+    }
+};
+
+
+TEST(serial, init)
+{
```
Comment (forGGe) [inline dev/bus/tests/serial_unit.cpp]: Let's be a more TDD'ish - add more tests, specifically at corner cases when overflow will occur

File: dev/bus/export/dev/serial.hpp [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+template <class PBus, size_t buf_size>
+void serial<PBus, buf_size>::bus_handler(bus_channel ch,
+        bus_event type, size_t total)
+{
+    if (ch == bus_channel::rx) {
+        ecl_assert(type == bus_event::tc);
+        m_rx_write_iter++;
+        ecl_assert(m_rx_write_iter == total);
+
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Overflow flag must be raised somewhere here. Proposed condition:
```

# Last time buffer was full, and no one has read any byte from it.
if buffer_is_full() and read_iter == write_iter:
   set_overflow()
   # Advance read_iter, so overflow condition will stay there until 
   # somebody will read data. And read iterator will point to the start
   # of the old, overflowed chunk of data. This has to be discussed.
   read_iter++;

# Checks that buffer is about to be full
if read_iter > 1 && write_iter == read_iter - 1: 
   set_buffer_full()

write_iter++
```

It means that . On the next ISR call if no read

File: dev/bus/export/dev/serial.hpp (hunk 9 above)
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Better to check if `read_iter == buffer_size`

File: dev/bus/tests/serial_unit.cpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    // Fill internal rx buffer and verify that xfer starts in the bus_handler
+    for (size_t i = 0; i < serial_t::buffer_size-1; i++) {
+        platform_mock::invoke(ecl::bus_channel::rx, ecl::bus_event::tc, i+1);
+    }
+
+    mock("platform_bus")
+        .expectOneCall("do_xfer")
+        .andReturnValue(static_cast<int>(ecl::err::ok));
+    platform_mock::invoke(ecl::bus_channel::rx, ecl::bus_event::tc, serial_t::buffer_size);
+}
```
Comment (forGGe) [inline dev/bus/tests/serial_unit.cpp]: You forgot to use `mock().checkExpectation()` probably

Comment (RostakaGmfun) [inline dev/bus/tests/serial_unit.cpp]: Done

File: dev/bus/tests/serial_unit.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+            CHECK_EQUAL(recv_returned, false);
+            platform_mock::invoke(ecl::bus_channel::rx, ecl::bus_event::tc, 1); // Unblock
+        });
+    auto expected = ecl::err::ok;
+    uint8_t byte;
+    recv_status.signal();
+    auto actual = serial_t::recv_byte(byte);
+    recv_returned = true;
+    t.join();
+    CHECK_EQUAL(expected, actual);
```
Comment (forGGe) [inline dev/bus/tests/serial_unit.cpp]: As a note for our offline discussion: let's check that byte received is the same byte that were written by `platform_mock`

File: dev/bus/tests/serial_unit.cpp [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+// TODO: move it to 'utils' headers and protect with check of
+// current test state (enabled or disabled)
+static SimpleString StringFrom(ecl::err err)
+{
+    return SimpleString{ecl::err_to_str(err)};
+}
+
+using serial_t = ecl::serial<platform_mock>;
+
+TEST_GROUP(serial)
```
Comment (forGGe) [inline dev/bus/tests/serial_unit.cpp]: I propose to add following tests (or you can implement some kind of variations if you will):

 1. Multiple `recv_byte` with total count less than internal buffer size, provided that there is enough data in the internal buffer. I.e. `recv_byte` will not block. 
 1. Multiple `recv_byte` with total count less than internal buffer size, provided that there is not enough data in the internal buffer to satisfy all `recv_byte` calls.. `recv_byte` will block when buffer will be empty. Thread that emulates ISR will trigger event with new data after `recv_byte` will block (adding simple delay between thread start and event firing should do the trick).
 1. Multiple `recv_byte` with total count bigger than internal buffer size,  provided that the internal buffer is full. Case is almost the same as previous one, however it will also test wraparound.

File: dev/bus/tests/serial_unit.cpp (hunk 14 above)
Comment (RostakaGmfun) [inline dev/bus/tests/serial_unit.cpp]: Done

File: dev/bus/export/dev/serial.hpp [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+{
+    ecl_assert(m_is_inited);
+    err result = err::ok;
+
+    // Restrict starting new xfer from bus_handler()
+    // to prevent data loss
+    m_rx_new_xfer_allowed = false;
+
+    // Wait until data is available on the read end
+    while ((m_rx_read_iter == m_rx_write_iter && m_rx_buffer_empty)) { }
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Let's use binary semaphore here.

File: dev/bus/export/dev/serial.hpp [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+{
+    ecl_assert(m_is_inited);
+    err result = err::ok;
+
+    // Restrict starting new xfer from bus_handler()
+    // to prevent data loss
+    m_rx_new_xfer_allowed = false;
+
+    // Wait until data is available on the read end
+    m_rx_is_data_available.wait();
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Let's do a small optimization:
```
if (m_rx_read_iter == m_rx_write_iter) {
     m_rx_is_data_available.wait();
}
```

File: dev/bus/export/dev/serial.hpp [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+        result = err::nobufs;
+    }
+
+    // Wrap rx read iterator
+    if (m_rx_read_iter == buffer_size) {
+        m_rx_read_iter = 0;
+    }
+
+    if (m_rx_read_iter != m_rx_write_iter) {
+        m_rx_is_data_available.signal();
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Not needed in case of optimization above

File: dev/bus/export/dev/serial.hpp [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+        // and we properly adjust the position of read iterator to be
+        // always ahead of the write iterator.
+        if (m_rx_read_iter == m_rx_write_iter) {
+            m_rx_read_iter++;
+            if (m_rx_read_iter == buffer_size) {
+                m_rx_read_iter = 0;
+            }
+        }
+
+        m_rx_is_data_available.signal();
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Let's call it only if there were no data before .
It possibly should be checked somewhere higher. Taking to the account our invariant that `m_rx_read_iter == m_rx_write_iter` only in case if there is no data left, we can signal semaphore using that check.

Avoiding excessive calls to `signal()` will prevent additional SVC calls and interrupts (in case RTOS is enabled)

File: dev/bus/export/dev/serial.hpp [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+    if (ch == bus_channel::rx) {
+        ecl_assert(type == bus_event::tc);
+        m_rx_write_iter++;
+        ecl_assert(m_rx_write_iter == total);
+
+        // The buffer has been overflowed.
+        // Move read iterator to point to the most relevant data
+        //
+        // It might be a bit unclear why the same expression
+        // m_rx_read_iter == m_rx_write_iter both means that the buffer is
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Hm, I suggest to somehow rewrite this statement. Actually, for any point of time if there is some data, read and write iters should not be equal. Only temporarily, within this frame and after write iterator increment they actually can overlap.

E.g.  there is no `both means`, rather `exception` in this particular place.

**Review by forGGe — COMMENTED** [review]
Some advices

File: platform/stm32/export/aux/usart_bus.hpp [hunk 21]
```diff
@@ -375,7 +399,6 @@ void usart_bus<dev>::reset_handler()
 template<usart_device dev>
 ecl::err usart_bus<dev>::do_xfer()
```
Comment (forGGe) [inline platform/stm32/export/aux/usart_bus.hpp]: Can be also simplified to just `do_tx(); do_rx()`

**Review by forGGe — COMMENTED** [review]
Small suggestion

**Comment by RostakaGmfun** [comment]
@forGGe, review, please

File: dev/bus/export/dev/serial.hpp [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+    static err send_buf(const uint8_t *buf, size_t &sz);
+
+private:
+    static void bus_handler(bus_channel ch, bus_event type, size_t total);
+
+    static bool m_is_inited;
+    static std::atomic_bool m_new_xfer_allowed;
+
+    static uint8_t m_rx_buffer[buf_size];
+    static std::atomic_size_t m_rx_write_iter; //!< Used for accepting data from the bus
```
Comment (forGGe) [inline dev/bus/export/dev/serial.hpp]: Consider this advice. Since you have 'invented' here smth like relaxed-lock queue, could you please add here a bit more info for future generations about domain, operation allowance and indicative conditions (predicates) that are associated with every iterator and flag?

For example, 
```
m_rx_write_iter 
constraints:
 - read only in normal (_not_ interrupt) execution mode, 
 - write/read in interrupt mode
 - m_rx_write_iter < buf_size is always true 
semantics:
 - denotes part of buffer that already populated with data
 - incremented with every byte reception in the event handler
indicative conditions:
 - in the normal execution mode m_rx_write_iter == m_rx_read_iter evaluates to true for case when buffer is empty, false otherwise 
 - in the interrupt execution mode, after byte reception and m_rx_write_iter increment m_rx_write_iter == m_rx_read_iter evaluates to true in case of overflow.
```

This is just rough example, pick your words and commenting approach. 

Still, it is just the advice, feel free to ignore it if you will consider it unnecessary (especially in the light that this info duplicates comments in event handler and recv/send methods)

**Review by forGGe — COMMENTED** [review]
Few notes

File: dev/bus/export/dev/serial.hpp [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+
+    //! Sends byte to a serial device.
+    //! \pre Driver is initialized.
+    //! \details It may not block if internal buffering is applied.
+    //! It may also block if platform device is not yet ready to
+    //! send data.
+    //! \param[in] byte Byte to send.
+    //! \return Operation status.
+    static err send_byte(uint8_t byte);
+
```
Comment (rasendubi) [inline dev/bus/export/dev/serial.hpp]: Code style nazi :D

Comment (RostakaGmfun) [inline dev/bus/export/dev/serial.hpp]: Done :sunglasses:

File: lib/utils/export/ecl/utils.hpp [hunk 24]
```diff
@@ -46,30 +46,44 @@ class safe_storage
...
     //! \tparam Args Types of the constructor args.
     //! \param  args T's ctor arguments.
-    //! \pre  Storage not yet initialized, i.e. ctor wasn't called yet.
+    //! \pre  Storage not yet initialized, i.e. ctor wasn't called yet
+    //!       or storage was deinitialized earlier.
     //! \post Storage initialized and object is ready to work.
     //! \note Violating pre-conditions leads to undefined behaviour.
     template<class ...Args>
-    static void init(Args&& ...args)
+    void init(Args&& ...args)
```
Comment (forGGe) [inline lib/utils/export/ecl/utils.hpp]: Wow! good catch!

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 065f1b9b514f88eac210b51e217abd3b787bea69
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/forGGe/theCore

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 065f1b9b514f88eac210b51e217abd3b787bea69
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/forGGe/theCore

---
This message was auto-generated by https://gitcop.com

**Comment by forGGe** [comment]
Development continuing in #310
